rust/extns/numaflow-nats/src/nats.rs

```diff
@@ -0,0 +1,277 @@
+    },
+}
+
+/// NatsActor represents a NATS subscriber actor
+struct NatsActor {
+    sub: async_nats::Subscriber,
+    read_timeout: Duration,
+    batch_size: usize,
+    handler_rx: mpsc::Receiver<NatsActorMessage>,
+}
+
+impl NatsActor {
+    async fn start(
+        config: NatsSourceConfig,
+        batch_size: usize,
+        read_timeout: Duration,
+        handler_rx: mpsc::Receiver<NatsActorMessage>,
+    ) -> Result<()> {
+        let mut conn_opts = make_conn_opts();
+        if let Some(auth) = config.auth {
+            conn_opts = match auth {
+                NatsAuth::Basic { username, password } => {
+                    conn_opts.user_and_password(username, password)
+                }
+                NatsAuth::NKey(nkey) => conn_opts.nkey(nkey),
+                NatsAuth::Token(token) => conn_opts.token(token),
+            };
+        }
+        if let Some(tls_config) = config.tls {
+            conn_opts = tls::configure_tls(conn_opts, tls_config)?;
+        }
+        let client = async_nats::connect_with_options(&config.addr, conn_opts)
+            .await
+            .map_err(|err| Error::Connection {
+                server: config.addr.to_string(),
+                error: err.to_string(),
+            })?;
+
+        // Wait for the subscription to be ready
+        let sub = client
+            .queue_subscribe(config.subject.clone(), config.queue.clone())
+            .await
+            .map_err(|err| Error::Subscription {
+                subject: config.subject.clone(),
+                queue: config.queue.clone(),
+                error: err.to_string(),
+            })?;
+
+        // Spawn the NATS actor
+        tokio::spawn(async move {
+            let mut actor = NatsActor {
+                sub,
+                read_timeout,
+                batch_size,
+                handler_rx,
+            };
+            tracing::info!(subject=?config.subject, queue=?config.queue, "Starting NATS source actor...");
+            actor.run().await;
+        });
+
+        Ok(())
+    }
+
+    async fn run(&mut self) {
+        while let Some(msg) = self.handler_rx.recv().await {
+            self.handle_message(msg).await;
+        }
+    }
+
+    /// Reads messages from the NATS subscriber, up to batch_size or until timeout
+    async fn read_messages(&mut self) -> Result<Vec<NatsMessage>> {
+        let mut messages: Vec<NatsMessage> = Vec::with_capacity(self.batch_size);
+        let timeout = tokio::time::timeout(self.read_timeout, std::future::pending::<()>());
+        tokio::pin!(timeout);
+        loop {
+            if messages.len() >= self.batch_size {
+                break;
+            }
+            tokio::select! {
+                biased;
+
+                _ = &mut timeout => {
+                    tracing::debug!(msg_count = messages.len(), "Timed out waiting for NATS messages");
+                    break;
+                }
+                maybe_msg = self.sub.next() => {
```

Comment:
should we do batch read? even in https://github.com/numaproj/numaflow/blob/3a82110b1ad4a5e50e1bb5bddfdfeb7e7c9de43f/rust/extns/numaflow-nats/src/jetstream.rs#L243 this could cause duplicates if there is backpressure and writes and not going through. 

This is because async-nats lib is will read ahead and the ack timeout will be triggered causing replay.